[Proposal]: Global Package Reference for Central Package Management (#11794)
NuGet/Home: proposed/2022/globalpackagereference.md

every assembly in your repository.  Instead, individual projects should express their build time dependencies and project references should receive
them transitively.  But you should not need to add a package reference in order to have your whole tree use code analysis or assembly versioning.

## Explanation

### Functional explanation
<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
We'll introduce a new MSBuild item group named `<GlobalPackageReference />` which will define packages that every project should reference but has
default metadata to ensure that only package assets related to development purposes are used.

**NOTE:** This functionality will only be enabled if a user has opt-ed into [Central Package Management (CPM)](https://docs.microsoft.com/nuget/consume-packages/central-package-management).

```xml
<Project>
  <ItemGroup>

>     <GlobalPackageReference Include="Nerdbank.GitVersioning" Version="1.0.0" />

Reviewer: I don't think the name `GlobalPackageReference` is correct. 

The `Global` part is a function of where this is written, not the value of the item itself. 

The itself is what was said above...a development dependency reference. 

So if we were to add this functionality, the name should be changed.

Reviewer 2: Naming suggestion: `CentralPackageReference`, to reflect Central Package Management usage.

Reviewer 2: Naming suggestion: `CentralDevPackageReference`, toe reflect CPM usage on a development dependency.

Reviewer 3: I would suggest `BuildOnlyPackageReference` or `DevOnlyPackageReference` but I suspect `GlobalPackageReference` will be an appropriate naming considering the purpose of it. Also `CentralPackageReference` doesn't sound right compared with the former. I can also suggest `SharedPackageReference` and `CommonPackageReference` but @jeffkl would've already pondered about it when he named it. So, it's better to ask him.

Reviewer 3: I would suggest `BuildOnlyPackageReference` or `DevOnlyPackageReference` but I suspect `GlobalPackageReference` will be an appropriate name considering the location and the **supposed** purpose of it. The default is development only dependencies but it's not stopping us from including other assets.

Also `CentralPackageReference` doesn't sound right compared with the former. I can also suggest `SharedPackageReference` and `CommonPackageReference` but @jeffkl would've already pondered about it when he named it. So, it's up to him.

Author: After two months of collecting feedback, `GlobalPackageReference` appears to be the most voted for naming based on internal and external surveys.

https://twitter.com/nuget/status/1545037921455153155

I think we should move forward with `GlobalPackageReference` at this point.

>     <GlobalPackageReference Include="Nerdbank.GitVersioning" Version="1.0.0" />

Reviewer 4: If CPM is a requirement, then shouldn't we not put a version here and put that in the Directory.Packages.props? Or are we trying to put a harder line between the two types of packages?

Author: The idea is that `Directory.Packages.props` contains all of your package versions and `GlobalPackageReference` items should really only be declared in this file.  The documentation will try to make this clear, `GlobalPackageReference` items should only be in `Directory.Packages.props` and not in individual projects themselves.

  </ItemGroup>
</Project>
```

### Technical explanation
<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
`<GlobalPackageReference />` will just be syntactical sugar for a `<PackageReference />` with the following metadata:

> * IncludeAssets = `Analyzers;Build;BuildMultitargeting;BuildTransitive`

Reviewer: Note that this is *different* from how `DevelopmentDependency` works in PackageReference.
If you read that thread, you'll discover that the feature is a `personal favorite` of mine :D 

If I could break things, I'd say development dependencies:

* Can't bring compile/runtime assets
* Have to be self contained, and can't have dependencies
* Are not part of the standard project graph, something theoretically addressed by not allowing them to have dependencies.

Note that what I described is effectively the guidance for writing msbuild task packages.

Reviewer: See details: https://github.com/NuGet/Home/pull/11794

Author: I'm going to remove all references to "development dependency" so there is no confusing.  This feature is about making sure repo-wide packages only consume _build_ assets.

Author: Based on our conversation we discovered that there are a lot of packages labeled as "development dependencies" with runtime assets.  So I've adjusted the default IncludeAssets to match that.  Otherwise a package with a runtime asset would behave differently as a "global" package reference vs normal.

> * PrivateAssets = `All`

Reviewer 3: ```suggestion
* IncludeAssets = `Runtime;Build;Native;ContentFiles;Analyzers`
* ExcludeAssets = `Compile`
* PrivateAssets = `All`
```

Author: We've decided to align this more with the default development dependency assets `runtime; build; native; contentfiles; analyzers` so that package references behave the same in both scenarios.

Reviewer: The Pack metadata doesn't apply to Package/ProjectReferences anyways. 

It's only a thing in some content items.

    <GlobalPackageReference>
      <!--
        Default global package references to only consume the Analyzers and Build logic in a package.
        This helps ensure that the package assets are not passed to the compiler or copied to the
        output directory.  Having a compile-time reference to a package for all projects in a tree
        is not recommended.  You should only have "global" references to packages that are used for
        build.
      -->
      <IncludeAssets Condition="'$(RestoreEnableGlobalPackageReferenceIncludeAssetsBuildAnalyzers)' != 'false'">Analyzers;Build;BuildMultitargeting;BuildTransitive</IncludeAssets>
      <!--
        Default global package references to have all assets private.  This is because global package
        references are generally stuff like versioning, signing, etc and should not flow to downstream
        dependencies.  Also, global package references are already referenced by every project in the
        tree so we don't need them to be transitive.
      -->
      <PrivateAssets Condition="'$(RestoreEnableGlobalPackageReferencePrivateAssetsAll)' != 'false'">All</PrivateAssets>
    </GlobalPackageReference>

>   </ItemDefinitionGroup>

Reviewer 3: ```suggestion
  <ItemDefinitionGroup Condition="'$(UseGlobalPackageReferences)' == 'true'">
    <GlobalPackageReference>
      <!--
        Default global package references should only consume Development Dependency assets.
        This helps to ensure that other package assets are not passed to the compiler or copied to output.
        Having a compile-time reference to a package for all projects in a tree is not recommended.
        You should only have "global" references to packages that are used only in build-time.
      -->
      <IncludeAssets Condition="'$(EnableGlobalPackageReferenceDefaultIncludeAssets)' != 'false'">Analyzers;Build;ContentFiles;Native;Runtime</IncludeAssets>
      <!-- These assets are not recommended to be required at all, so they are excluded explicitly -->
      <ExcludeAssets Condition="'$(EnableGlobalPackageReferenceDefaultExcludeAssets)' != 'false'">Compile</ExcludeAssets>
      <!--
        Default global package references should have all assets private to the project.
        This is because "global" package references are generally stuff like versioning, signing, etc and
        they should not flow to downstream dependencies. Also, they are already referenced by every
        project in the tree; so, we don't need them to be transitive.
      -->
      <PrivateAssets Condition="'$(EnableGlobalPackageReferenceDefaultPrivateAssets)' != 'false'">All</PrivateAssets>
    </GlobalPackageReference>
  </ItemDefinitionGroup>
```

Author: Since the packages are PrivateAssets=All they won't participate in pack.  They'll participate in Publish if they have any runtime or contentfiles assets but that's expected.

      Global package references have all assets private because central package references are generally packages that provide
      versioning, signing, etc and should not flow to downstream dependencies.  Also, central package references are already
      referenced by every project in the tree so they don't need to be transitive.
    -->
    <PackageReference Include="@(GlobalPackageReference)"
                      Version=""
                      IncludeAssets="runtime; build; native; contentfiles; analyzers"

>                       PrivateAssets="All" />

Reviewer 3: With the item definition, you can override the GPC's metadata but this removes the ability to that. If we specify GPC with a custom metadata, those will get overriden here.

> | `RestoreEnableGlobalPackageReferencePrivateAssetsAll` | Enables or disables defaulting the `PrivateAssets` metadata for global package references | `true`|

Reviewer 3: I'm suggesting a new property to oversee all CPM-related features. i.e.: `ManagePackagesCentrally` though naming is up for debate.

```suggestion
| Property Name | Function | Default value |
|---|---|---|
| `ManagePackagesCentrally` | Enables or disables central package management and all dependent features | `false` |
| `ManagePackageVersionsCentrally` | Enables or disables managing of package versions in a centralized way | `$(ManagePackagesCentrally)` or `false` |
| `UseGlobalPackageReferences` | Enables or disables global package references | `$(ManagePackagesCentrally)` or `true` |
| `EnableGlobalPackageReferenceDefaultIncludeAssets` | Enables or disables defaulting the `IncludeAssets` metadata for global package references | **nil** |
| `EnableGlobalPackageReferenceDefaultExcludeAssets` | Enables or disables defaulting the `ExcludeAssets` metadata for global package references | **nil** |
| `EnableGlobalPackageReferenceDefaultPrivateAssets` | Enables or disables defaulting the `PrivateAssets` metadata for global package references | **nil** |
```

#### Other names:

* `EnableDefaultPrivateAssetsInGlobalPackageReference`
* `EnableDefaultPrivateAssetsForGlobalPackageReference`

The above default value implementation expressed in targets:

```xml
  <PropertyGroup>
    <UseGlobalPackageReferences Condition="'$(UseGlobalPackageReferences)' == ''">$([MSBuild]::ValueOrDefault('$(ManagePackagesCentrally)', 'true'))</UseGlobalPackageReferences>
    <ManagePackageVersionsCentrally Condition="'$(ManagePackageVersionsCentrally)' == ''">$([MSBuild]::ValueOrDefault('$(ManagePackagesCentrally)', 'false'))</ManagePackageVersionsCentrally>
  </PropertyGroup>
```

Reviewer 3: Also, instead of opting-in, just as I mentioned detecting `Directory.Packages.props` to enable central package versions, do the same for "Global" package references. We could do this either by detecting "packages" props or by detecting "build" props and non-empty **GPR** items check.

`GlobalPackageReference` items will then be copied to the `PackageReference` item group along with their metadata:
```xml

> <ItemGroup>
>   <PackageReference Include="@(GlobalPackageReference)" Version="" />
>   <PackageVersion Include="@(GlobalPackageReference)" Version="%(Version)" />
> </ItemGroup>

Reviewer 3: ```suggestion
<ItemGroup>
  <PackageReference Include="@(GlobalPackageReference)" Condition="'$(ManagePackageVersionsCentrally)' != 'true'" />
  <PackageReference Include="@(GlobalPackageReference)" Version="" Condition="'$(ManagePackageVersionsCentrally)' == 'true'" />
</ItemGroup>

<ItemGroup>
  <PackageVersion Include="@(GlobalPackageReference->ClearMetadata())" Version="%(Version)" />
</ItemGroup>
```

Reviewer 3: In response to <https://github.com/NuGet/Home/pull/11794#issuecomment-1161166465>
